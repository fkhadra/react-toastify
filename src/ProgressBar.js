import React from 'react';
import PropTypes from 'prop-types';
import { css } from "glamor";

import { TYPE } from './constant';
import style from "./style";

const trackProgress = css.keyframes('track-progress', {
  "0%": { width: "100%" },
  "100%": { width: 0 }
});

const progress = (type, isRunning, hide, delay) => css({
  position: "absolute",
  bottom: 0,
  left: 0,
  width: 0,
  height: "5px",
  zIndex: style.zIndex,
  opacity: hide ? 0: 0.7,
  animation: `${trackProgress} linear 1`,
  animationPlayState: isRunning ? 'running' : 'paused',
  animationDuration: `${delay}ms`,
  backgroundColor: "rgba(255,255,255,.7)",
  ...type === "default" ? { background: style.colorProgressDefault } : {}
});

function ProgressBar({ delay, isRunning, closeToast, type, hide, className }) {
  return (
    <div
      {...progress(type, isRunning, hide, delay)}
      className={className}
      onAnimationEnd={closeToast}
    />
  );
}

ProgressBar.propTypes = {
  /**
   * The animation delay which determine when to close the toast
   */
  delay: PropTypes.number.isRequired,

  /**
   * Whether or not the animation is running or paused
   */
  isRunning: PropTypes.bool.isRequired,

  /**
   * Func to close the current toast
   */
  closeToast: PropTypes.func.isRequired,

  /**
   * Optional type : info, success ...
   */
  type: PropTypes.string,

  /**
   * Hide or not the progress bar
   */
  hide: PropTypes.bool,

  /**
   * Optionnal className
   */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
};

ProgressBar.defaultProps = {
  type: TYPE.DEFAULT,
  hide: false,
  className: ''
};

export default ProgressBar;
