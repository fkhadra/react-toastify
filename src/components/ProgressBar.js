import { __assign } from "tslib";
import React from 'react';
import cx from 'clsx';
import { isFn } from './../utils';
export function ProgressBar(_a) {
    var _b, _c;
    var delay = _a.delay, isRunning = _a.isRunning, closeToast = _a.closeToast, _d = _a.type, type = _d === void 0 ? "default" /* Type.DEFAULT */ : _d, hide = _a.hide, className = _a.className, userStyle = _a.style, controlledProgress = _a.controlledProgress, progress = _a.progress, rtl = _a.rtl, isIn = _a.isIn, theme = _a.theme;
    var isHidden = hide || (controlledProgress && progress === 0);
    var style = __assign(__assign({}, userStyle), { animationDuration: "".concat(delay, "ms"), animationPlayState: isRunning ? 'running' : 'paused' });
    if (controlledProgress)
        style.transform = "scaleX(".concat(progress, ")");
    var defaultClassName = cx("".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar"), controlledProgress
        ? "".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--controlled")
        : "".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--animated"), "".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar-theme--").concat(theme), "".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--").concat(type), (_b = {},
        _b["".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--rtl")] = rtl,
        _b));
    var classNames = isFn(className)
        ? className({
            rtl: rtl,
            type: type,
            defaultClassName: defaultClassName
        })
        : cx(defaultClassName, className);
    // 🧐 controlledProgress is derived from progress
    // so if controlledProgress is set
    // it means that this is also the case for progress
    var animationEvent = (_c = {},
        _c[controlledProgress && progress >= 1
            ? 'onTransitionEnd'
            : 'onAnimationEnd'] = controlledProgress && progress < 1
            ? null
            : function () {
                isIn && closeToast();
            },
        _c);
    // TODO: add aria-valuenow, aria-valuemax, aria-valuemin
    return (React.createElement("div", { className: "".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--wrp"), "data-hidden": isHidden },
        React.createElement("div", { className: "".concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--bg ").concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar-theme--").concat(theme, " ").concat("Toastify" /* Default.CSS_NAMESPACE */, "__progress-bar--").concat(type) }),
        React.createElement("div", __assign({ role: "progressbar", "aria-hidden": isHidden ? 'true' : 'false', "aria-label": "notification timer", className: classNames, style: style }, animationEvent))));
}
//# sourceMappingURL=ProgressBar.js.map