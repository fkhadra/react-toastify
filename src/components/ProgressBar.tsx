import React from 'react';
import cx from 'clsx';

import { Default, isFn, Type } from './../utils';
import { TypeOptions, ToastClassName, Theme } from '../types';

export interface ProgressBarProps {
  /**
   * The animation delay which determine when to close the toast
   */
  delay: number;

  /**
   * Whether or not the animation is running or paused
   */
  isRunning: boolean;

  /**
   * Func to close the current toast
   */
  closeToast: () => void;

  /**
   * Optional type : info, success ...
   */
  type?: TypeOptions;

  /**
   * The theme that is currently used
   */
  theme: Theme;

  /**
   * Hide or not the progress bar
   */
  hide?: boolean;

  /**
   * Optionnal className
   */
  className?: ToastClassName;

  /**
   * Optionnal inline style
   */
  style?: React.CSSProperties;

  /**
   * Tell wether or not controlled progress bar is used
   */
  controlledProgress?: boolean;

  /**
   * Controlled progress value
   */
  progress?: number | string;

  /**
   * Support rtl content
   */
  rtl?: boolean;

  /**
   * Tell if the component is visible on screen or not
   */
  isIn?: boolean;
}

export function ProgressBar({
  delay,
  isRunning,
  closeToast,
  type = Type.DEFAULT,
  hide,
  className,
  style: userStyle,
  controlledProgress,
  progress,
  rtl,
  isIn,
  theme
}: ProgressBarProps) {
  const isHidden = hide || (controlledProgress && progress === 0);
  const style: React.CSSProperties = {
    ...userStyle,
    animationDuration: `${delay}ms`,
    animationPlayState: isRunning ? 'running' : 'paused',
    opacity: isHidden ? 0 : 1
  };

  if (controlledProgress) style.transform = `scaleX(${progress})`;
  const defaultClassName = cx(
    `${Default.CSS_NAMESPACE}__progress-bar`,
    controlledProgress
      ? `${Default.CSS_NAMESPACE}__progress-bar--controlled`
      : `${Default.CSS_NAMESPACE}__progress-bar--animated`,
    `${Default.CSS_NAMESPACE}__progress-bar-theme--${theme}`,
    `${Default.CSS_NAMESPACE}__progress-bar--${type}`,
    {
      [`${Default.CSS_NAMESPACE}__progress-bar--rtl`]: rtl
    }
  );
  const classNames = isFn(className)
    ? className({
        rtl,
        type,
        defaultClassName
      })
    : cx(defaultClassName, className);

  // 🧐 controlledProgress is derived from progress
  // so if controlledProgress is set
  // it means that this is also the case for progress
  const animationEvent = {
    [controlledProgress && (progress as number)! >= 1
      ? 'onTransitionEnd'
      : 'onAnimationEnd']:
      controlledProgress && (progress as number)! < 1
        ? null
        : () => {
            isIn && closeToast();
          }
  };

  // TODO: add aria-valuenow, aria-valuemax, aria-valuemin

  return (
    <div
      role="progressbar"
      aria-hidden={isHidden ? 'true' : 'false'}
      aria-label="notification timer"
      className={classNames}
      style={style}
      {...animationEvent}
    />
  );
}
