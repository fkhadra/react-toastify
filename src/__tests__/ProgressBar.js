/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'

import ProgressBar from './../ProgressBar';

const REQUIRED_PROPS = {
  delay: 5000,
  isRunning: true,
  closeToast: jest.fn()
};

describe('ProgressBar', () => {
  it('Should merge className', () => {
    const component = shallow(
      <ProgressBar
        {...REQUIRED_PROPS}
        className="test"
      />
    );
    expect(component.find('.test')).toHaveLength(1);
  });

  it('Should call closeToast function when animation end', () => {
    const component = shallow(<ProgressBar {...REQUIRED_PROPS} />);

    expect(REQUIRED_PROPS.closeToast).not.toHaveBeenCalled();
    component.simulate('animationEnd');
    expect(REQUIRED_PROPS.closeToast).toHaveBeenCalled();
  });

  it("Should be able to hide the progress bar", () => {
    const component = shallow(<ProgressBar {...REQUIRED_PROPS} hide/>);
    expect(toJson(component)).toMatchSnapshot();
  });
  
  it("Should be able to pause animation", () => {
    const component = shallow(<ProgressBar {...REQUIRED_PROPS} isRunning={false} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
