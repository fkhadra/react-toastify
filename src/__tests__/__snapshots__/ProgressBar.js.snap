// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Should be able to hide the progress bar 1`] = `
.css-qkhwto,
[data-css-qkhwto] {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0;
  height: 5px;
  z-index: 9999;
  opacity: 0;
  animation: track-progress_1pd4w4j linear 1;
  animation-play-state: running;
  animation-duration: 5000ms;
  background-color: rgba(255,255,255,.7);
  background: -webkit-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: -moz-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  -webkit-animation: track-progress_1pd4w4j linear 1;
  -webkit-animation-play-state: running;
  -webkit-animation-duration: 5000ms;
}

<div
  className=""
  data-css-qkhwto=""
  onAnimationEnd={[Function]}
/>
`;

exports[`ProgressBar Should be able to pause animation 1`] = `
.css-1tcv8a9,
[data-css-1tcv8a9] {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0;
  height: 5px;
  z-index: 9999;
  opacity: 0.7;
  animation: track-progress_1pd4w4j linear 1;
  animation-play-state: paused;
  animation-duration: 5000ms;
  background-color: rgba(255,255,255,.7);
  background: -webkit-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: -moz-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  -webkit-animation: track-progress_1pd4w4j linear 1;
  -webkit-animation-play-state: paused;
  -webkit-animation-duration: 5000ms;
}

<div
  className=""
  data-css-1tcv8a9=""
  onAnimationEnd={[Function]}
/>
`;
